internal/mempool/v1/mempool.go

```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
+		"inserted new valid transaction",
+		"priority", wtx.Priority(),
 		"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
 		"height", txmp.height,
 		"num_txs", txmp.Size(),
 	)
 	txmp.notifyTxsAvailable()
-
 }
 
-// defaultTxCallback performs the default CheckTx application callback. This is
-// NOT executed when a transaction is first seen/received. Instead, this callback
-// is executed during re-checking transactions (if enabled). A caller, i.e a
-// block proposer, acquires a mempool write-lock via Lock() and when executing
-// Update(), if the mempool is non-empty and Recheck is enabled, then all
-// remaining transactions will be rechecked via CheckTxAsync. The order in which
-// they are rechecked must be the same order in which this callback is called
-// per transaction.
-func (txmp *TxMempool) defaultTxCallback(req *abci.Request, res *abci.Response) {
-	if txmp.recheckCursor == nil {
-		return
+func (txmp *TxMempool) insertTx(wtx *WrappedTx) {
+	elt := txmp.txs.PushBack(wtx)
+	txmp.txByKey[wtx.tx.Key()] = elt
+	if s := wtx.Sender(); s != "" {
+		txmp.txBySender[s] = elt
 	}
 
-	txmp.metrics.RecheckTimes.Add(1)
+	atomic.AddInt64(&txmp.txsBytes, wtx.Size())
+}
+
+// recheckTxCallback handles the responses from ABCI CheckTx calls issued
+// during the recheck phase of a block Update. It updates the recheck counter
+// and removes any transactions invalidated by the application.
+//
+// This callback is NOT executed for the initial CheckTx on a new transaction;
+// that case is handled by initTxCallback instead.
+func (txmp *TxMempool) recheckTxCallback(req *abci.Request, res *abci.Response) {
+	txmp.mtx.Lock()
+	defer txmp.mtx.Unlock()
 
 	checkTxRes, ok := res.Value.(*abci.Response_CheckTx)
 	if !ok {
-		txmp.logger.Error("received incorrect type in mempool callback",
+		txmp.logger.Error("mempool: received incorrect result type in CheckTx callback",
 			"expected", reflect.TypeOf(&abci.Response_CheckTx{}).Name(),
 			"got", reflect.TypeOf(res.Value).Name(),
 		)
 		return
 	}
-	tx := req.GetCheckTx().Tx
-	wtx := txmp.recheckCursor.Value.(*WrappedTx)
-
-	// Search through the remaining list of tx to recheck for a transaction that matches
-	// the one we received from the ABCI application.
-	for {
-		if bytes.Equal(tx, wtx.tx) {
-			// We've found a tx in the recheck list that matches the tx that we
-			// received from the ABCI application.
-			// Break, and use this transaction for further checks.
-			break
-		}
-
-		txmp.logger.Error(
-			"re-CheckTx transaction mismatch",
-			"got", wtx.tx.Hash(),
-			"expected", types.Tx(tx).Key(),
-		)
 
-		if txmp.recheckCursor == txmp.recheckEnd {
-			// we reached the end of the recheckTx list without finding a tx
-			// matching the one we received from the ABCI application.
-			// Return without processing any tx.
-			txmp.recheckCursor = nil
+	numLeft := atomic.AddInt64(&txmp.txRecheck, -1)
+	if numLeft <= 0 {
+		defer txmp.notifyTxsAvailable()
+		if numLeft < 0 {
 			return
 		}
```

Comment:
What does this do? I'm slightly confused

Reply:
I'll add a comment. It's detecting that we are finished with a recheck cycle. The extra check is because we might have been flushed during the recheck, so we will still receive the callbacks (so we don't stall the client) but we don't process the tx (which is, perforce, gone)